File: src/main/java/cn/solarmoon/idyllic_food_diary/element/matter/cookware/steamer/SteamerItem.java
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.steamer;

import cn.solarmoon.idyllic_food_diary.feature.generic_recipe.steaming.ISteamingRecipe;
import cn.solarmoon.idyllic_food_diary.registry.common.IMBlocks;
import net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.client.extensions.common.IClientItemExtensions;
import net.minecraftforge.items.ItemStackHandler;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class SteamerItem extends BlockItem {

    public SteamerItem() {
        super(IMBlocks.STEAMER.get(), new Properties().stacksTo(1));
    }

    @Override
    public void initializeClient(Consumer<IClientItemExtensions> consumer) {
        consumer.accept(new IClientItemExtensions() {
            @Override
            public BlockEntityWithoutLevelRenderer getCustomRenderer() {
                return new SteamerItemRenderer();
            }
        });
    }

    public static List<ItemStackHandler> getInvFromItem(ItemStack steamerItemStack) {
        List<ItemStackHandler> invList = new ArrayList<>();
        CompoundTag beT = BlockItem.getBlockEntityData(steamerItemStack);
        if (beT != null && beT.contains(ISteamingRecipe.STEAMER_INV_LIST)) {
            ListTag listTag = beT.getList(ISteamingRecipe.STEAMER_INV_LIST, ListTag.TAG_COMPOUND);
            for (int i = 0; i < listTag.size(); i++) {
                ItemStackHandler s = new ItemStackHandler();
                s.deserializeNBT(listTag.getCompound(i));
                invList.add(s);
            }
        }
        return invList;
    }

    public static int getStackAmount(ItemStack steamerItemStack) {
        CompoundTag beT = BlockItem.getBlockEntityData(steamerItemStack);
        if (beT != null && beT.contains(SteamerBlockEntity.STACK)) {
            return beT.getInt(SteamerBlockEntity.STACK);
        }
        return 0;
    }

    public static ItemStack getLid(ItemStack steamerItemStack) {
        if (steamerItemStack.isEmpty()) return ItemStack.EMPTY;
        CompoundTag beT = BlockItem.getBlockEntityData(steamerItemStack);
        if (beT != null && beT.contains(SteamerBlockEntity.LID)) {
            return ItemStack.of(beT.getCompound(SteamerBlockEntity.LID));
        }
        return ItemStack.EMPTY;
    }

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/element/matter/cookware/steamer/SteamerItem.kt
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.steamer

import cn.solarmoon.idyllic_food_diary.registry.common.IFDBlockEntities
import cn.solarmoon.idyllic_food_diary.registry.common.IFDBlocks
import net.minecraft.core.HolderLookup
import net.minecraft.core.component.DataComponents
import net.minecraft.nbt.ListTag
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.component.CustomData
import net.minecraft.world.level.block.entity.BlockEntityType

class SteamerItem: BlockItem(IFDBlocks.STEAMER.get(), Properties().stacksTo(1)) {

    companion object {
        @JvmStatic
        fun singleEmptyLayer(registries: HolderLookup.Provider) = CustomData.EMPTY.update {
            val listTag = ListTag()
            listTag.add(SteamerInventory(4).serializeNBT(registries))
            it.putString("id", BlockEntityType.getKey(IFDBlockEntities.STEAMER.get()).toString()) // 此项为系统自动添加，不可能不加
            it.put(SteamerBlockEntity.INV_LIST, listTag)
            it.put(SteamerBlockEntity.LID, ItemStack.EMPTY.saveOptional(registries))
        }

        @JvmStatic
        fun getInvListTag(stack: ItemStack, registries: HolderLookup.Provider): ListTag {
            val tag = stack.getOrDefault(DataComponents.BLOCK_ENTITY_DATA, singleEmptyLayer(registries)).copyTag()
            return tag.getList(SteamerBlockEntity.INV_LIST, ListTag.TAG_COMPOUND.toInt())
        }

        @JvmStatic
        fun getLid(stack: ItemStack, registries: HolderLookup.Provider): ItemStack {
            val tag = stack.getOrDefault(DataComponents.BLOCK_ENTITY_DATA, singleEmptyLayer(registries)).copyTag()
            return ItemStack.parseOptional(registries, tag.getCompound(SteamerBlockEntity.LID))
        }

        @JvmStatic
        fun setLid(stack: ItemStack, lid: ItemStack, registries: HolderLookup.Provider) {
            stack.update(DataComponents.BLOCK_ENTITY_DATA, singleEmptyLayer(registries)) {
                it.update {
                    it.put(SteamerBlockEntity.LID, lid.saveOptional(registries))
                }
            }
        }
    }

}